roles/expirationd.lua

```diff
@@ -0,0 +1,302 @@
+    s = {type = "string", err = "a string"},
+    f = {type = "string", transform = load_function, err = "a function name in _G or in box.func"},
+    t = {type = "table", err = "a table"},
+    any = {err = "any type"},
+}
+
+local opts_map = {
+    args = {"any"},
+    atomic_iteration = {"b"},
+    force = {"b"},
+    force_allow_functional_index = {"b"},
+    full_scan_delay = {"n"},
+    full_scan_time = {"n"},
+    index = {"n", "s"},
+    iterate_with = {"f"},
+    iteration_delay = {"n"},
+    iterator_type = {"n", "s"},
+    on_full_scan_complete = {"f"},
+    on_full_scan_error = {"f"},
+    on_full_scan_start = {"f"},
+    on_full_scan_success = {"f"},
+    process_expired_tuple = {"f"},
+    process_while = {"f"},
+    start_key = {"f", "t"},
+    tuples_per_iteration = {"n"},
+    vinyl_assumed_space_len_factor = {"n"},
+    vinyl_assumed_space_len = {"n"},
+}
+
+local function get_param(param_name, value, types)
+    local found = false
+    for _, t in ipairs(types) do
+        local type_opts = types_map[t]
+        if type_opts == nil then
+            error(role_name .. ": unsupported type option")
+        end
+        if not type_opts.type or type(value) == type_opts.type then
+            if type_opts.transform then
+                local tmp = type_opts.transform(value)
+                if tmp then
+                    value = tmp
+                    found = true
+                    break
+                end
+            else
+                found = true
+                break
+            end
+        end
+    end
+
+    -- Small hack because in tarantool role we wait for functions to be created.
+    for _, t in ipairs(types) do
+        if not found and t == "f" then
+            return nil, true, nil
+        end
+    end
+
+    if not found then
+        local err = role_name .. ": " .. param_name .. " must be "
+        for i, t in ipairs(types) do
+            err = err .. types_map[t].err
+            if i ~= #types then
+                err = err .. " or "
+            end
+        end
+        return nil, false, err
+    end
+
+    return value, true, nil
+end
+
+local function get_task_options(opts)
+    if opts == nil then
+        return
+    end
+
+    local missed_functions = {}
+
+    for opt, val in pairs(opts) do
+        if type(opt) ~= "string" then
+            error(role_name .. ": an option must be a string")
+        end
+        if opts_map[opt] == nil then
+            error(role_name .. ": unsupported option '" .. opt .. "'")
+        end
+        local res, ok, err = get_param("options." .. opt, val, opts_map[opt])
+        if not ok then
+            error(err)
+        end
+        if ok and res == nil and opts_map[opt][1] == "f" then
+            table.insert(missed_functions, val)
+        end
+        opts[opt] = res
+    end
+
+    return opts, missed_functions
+end
+
+local function get_task_config(task_conf)
+    -- setmetatable resets __newindex write protection on a copy.
```

Comment:
Why do we need write permissions here?

Reply:
It is code used from cartridge role, so @oleg-jukovec  may answer better than me. I left it because w/o it role can not access config on https://github.com/tarantool/expirationd/blob/33a7d36b4e603dda20e101fea1ab2ee0ee85fea1/roles/expirationd.lua#L151-L154. I had not looked at the cause of this issue, just used existing solution.

Reply:
We make a copy of the configuration to make some checks/set defaults/etc and use it in our internal code. To do that we need to reset a write protection from the copy. We don't modify original `task_conf` here.